base/server/upgrade/10.10.0/02-FixMissingCertAndRequestData.py

```diff
@@ -0,0 +1,76 @@
+# Authors:
+#     Endi S. Dewata <edewata@redhat.com>
+#
+# Copyright Red Hat, Inc.
+#
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+from __future__ import absolute_import
+import logging
+
+import pki
+
+logger = logging.getLogger(__name__)
+
+
+class FixMissingCertAndRequestData(pki.server.upgrade.PKIServerUpgradeScriptlet):
+
+    def __init__(self):
+        super(FixMissingCertAndRequestData, self).__init__()
+        self.message = 'Fix missing SSL server and subsystem cert/request data'
+
+    def upgrade_instance(self, instance):
+
+        subsystems = instance.get_subsystems()
+
+        # there should be at least a source and a target
+        if len(subsystems) < 2:
+            return
+
+        logger.info('Finding a subsystem that has the cert/request data')
+        source = self.find_source_subsystem(subsystems)
+
+        # fix all subsystems other than the source
+        for subsystem in subsystems:
+
+            if subsystem.name == source.name:
+                continue
+
+            logger.info('Importing cert/request data into %s subsystem', subsystem.name)
+            self.backup(subsystem.cs_conf)
+
+            subsystem.config['%s.sslserver.cert' % subsystem.name] = \
+                source.config['%s.sslserver.cert' % source.name]
+
+            subsystem.config['%s.subsystem.cert' % subsystem.name] = \
+                source.config['%s.subsystem.cert' % source.name]
+
+            subsystem.config['%s.sslserver.certreq' % subsystem.name] = \
+                source.config['%s.sslserver.certreq' % source.name]
+
+            subsystem.config['%s.subsystem.certreq' % subsystem.name] = \
+                source.config['%s.subsystem.certreq' % source.name]
+
+            subsystem.save()
+
+    def find_source_subsystem(self, subsystems):
+
+        # check each subsystem
+        for subsystem in subsystems:
```

Comment:
There's actually no guarantee that all PKI servers will have a CA subsystem, and even if it does, there's no guarantee if the CA will have all the cert info. This loop is trying to find the first other subsystem that has all the info, so most likely it will stop after the first or second iteration. It's based on the existing code:
https://github.com/dogtagpki/pki/blob/master/base/server/python/pki/server/deployment/pkihelper.py#L2934-L2942

A better solution would be to remove these params from CS.cfg so we don't need to maintain them anymore, but that can be done separately later.

Reply:
AH! ok. Now i understand; CA and KRA can be installed on 2 different machines but be in the same Security Domain.

Hmmm. Now, that raises me another question: Let's say machine A has CA and machine B has KRA. Now, if the upgrade script is running in machine B, where will it read these values from?

Reply:
This particular regression presumably only happened to the subsystems added after the first one, so that situation should not happen, or is outside the scope of this upgrade script. As mentioned above, a better solution is to remove it completely from CS.cfg.

Reply:
OK.